### ISSUE TITLE
id attribute of spans gets lost

### ISSUE THREAD

**Issue by tolot27** [jgm/pandoc#4305]
Pandoc version 2.1.1

Steps to reproduce with file `comment_id.native`:
```
[Para [Str "I",Space,Str "want",Space,Span ("",["comment-start"],[("id","0"),("author","Jesse Rosenthal"),("date","2016-05-09T16:13:00Z")]) [Str "I",Space,Str "left",Space,Str "a",Space,Str "comment."],Str "some",Space,Str "text",Space,Str "to",Space,Str "have",Space,Str "a",Space,Str "comment",Space,Span ("",["comment-end"],[("id","0")]) [],Str "on",Space,Str "it."]]
```
Command: `pandoc -t markdown comment_id.native -o - | pandoc -t native`
Output:
```
[Para [Str "I",Space,Str "want",Space,Span ("0",["comment-start"],[("author","Jesse Rosenthal"),("date","2016-05-09T16:13:00Z")]) [Str "I",Space,Str "left",Space,Str "a",Space,Str "comment."],Str "some",Space,Str "text",Space,Str "to",Space,Str "have",Space,Str "a",SoftBreak,Str "comment",Space,Span ("0",["comment-end"],[]) [],Str "on",Space,Str "it."]]
```
Here, the value of the id attribute has moved to the first string of the span.

**Comment by jgm** [comment]
+++ Mathias Walter [Jan 25 18 14:00 ]:
>   Here, the value of the id attribute has moved to the first string of
>   the span.

Yes.  That is not a mistake; this is where we put the id
attribute in the AST.  Pandoc stores attributes in a 3-tuple:

    (identifier, list of classes, list of other key-value attributes)

Perhaps the docx reader should put the id here instead of
putting it in the key-value list.

(Yeah, it's not ideal to have a representation of attributes
that allows two different ways to say the same thing.  Hard
to change now without a lot of effort, though.)

**Comment by tolot27** [comment]
And how to get this id using lua?

**Comment by jgm** [comment]
Try el.identifier where el is your span element.

+++ Mathias Walter [Jan 25 18 11:22 ]:

**Comment by jkr** [comment]
@jgm - this actually creates a bug in the docx writer, which I found while setting up the new test architecture. Right now, it actually creates corrupt oxxml:  `<w:comment w:id="" w:id="1" w:author="Jesse Rosenthal" w:date="2016-05-09T16:13:00Z">`. Note the double ids.

What we need is a rule of precedence here: use `(ID,_,_)` over `(_,_,[("id", ID)])` if it exists, otherwise use the ones in the kvs? Or prefer kvs? Or just disallow kvs with names "id" and "class" altogether (i.e., ignore them in writing output)? 

The last one---disallowing---might actually be the neatest and most consistent. (Think of it like "shadows an existing binding" warnings in ghc.)

**Comment by jgm** [comment]
+++ Jesse Rosenthal [Jan 26 18 12:13 ]:
>   What we need is a rule of precedence here: use (ID,_,_) over
>   (_,_,[("id", ID)]) if it exists, otherwise use the ones in the kvs? Or
>   prefer kvs? Or just disallow kvs with names "id" and "class" altogether
>   (i.e., ignore them in writing output)?

I think this is the approach I use in some writers.  Anyway,
it makes sense to me.

**Comment by tolot27** [comment]
Is this fixed in pandoc 2.2?

**Comment by jgm** [comment]
I've lost track of what the bug is, exactly (if there is a bug).

The original example (native to native via markdown) isn't really a bug.  You're not meant to put an "id" attribute in the key/value attribute list; the id should go in the first position of the attribute triple.  Ideally this would be made impossible with types, but that would be a big change.

@jkr mentioned an issue having to do with the docx reader and writer.  I'm not sure if there's a bug there, but it looks like the docx reader currently uses "id" in the key/value list part of attributes. That should change.  It's a very simple thing:
```
-      let attr = ("", ["comment-start"], [("id", cmtId), ("author", author), ("date", date)])
+     let attr = (cmtId, ["comment-start"], [("author", author), ("date", date)])
```
```
-     let attr = ("", ["comment-end"], [("id", cmtId)])
+     let attr = (cmtId, ["comment-end"], [])
```
Presumably some changes in the tests would also be needed, as well as a change in the docx writer, which has been written to conform to the behavior of the docx reader:

``` haskell
inlineToOpenXML' _ (Span (ident,["comment-start"],kvs) ils) = do
  -- prefer the "id" in kvs, since that is the one produced by the docx
  -- reader.
  let ident' = fromMaybe ident (lookup "id" kvs)
      kvs' = filter (("id" /=) . fst) kvs
  modify $ \st -> st{ stComments = (("id",ident'):kvs', ils) : stComments st }
  return [ mknode "w:commentRangeStart" [("w:id", ident')] () ]
```

**Comment by tolot27** [comment]
Yes, it makes sense. Or maybe `comment-id` to have the same naming style. As an alternative, a rename of `comment-start` and `comment-end` to `commentStart` and `commentEnd` respectively.
